proposals/0000-local-import.rst

```diff
@@ -0,0 +1,159 @@
+
+::
+
+  module MyModule where
+
+  import Data.Set (Set)
+  import qualified Data.Set as Set
+  
+  foo :: (Set Int, Set String)
+  foo = (s1, fromList ["foo", "bar", "baz"])
+    where
+      import Set
+      s1 = fromList [1, 2, 3]
+
+::
+
+  module Main where
+  import System.Environment as Env
+   
+  main :: IO ()
+  main = do
+    args <- Env.getArgs
+    case args of
+      x:y:_ -> …
+      _ -> do
+        import System.Exit as Exit
+        Exit.{ exitWith $ ExitFailure (-1) }
+
+Motivation
+------------
+
+The motivations are the following:
+
+- Making import lists smaller and easier to maintain.
+- Making it easier to understand why a module is imported.
+- Making it easier to refactor a definition, along with its imports.
+- Making it more practical to use qualified imports, especially with operators.
+
+Import lists in Haskell can quickly become unwieldly. The recent proliferation of GHC proposals intending to simplify module imports is good evidence that this feeling is widely shared.
+
+Oftentimes, a module is only imported for a single symbol or two. The purpose of such imports is not obvious at first glance and can significantly clutter the import list. Moving those imports closer to the use-site would solve both of these problems by making the intent clearer and removing them from the toplevel list. Moreover, definitions would be easier to refactor, as they would be more self-contained ; for example, it would be easy to remove a definition along with its specific imports.
+
+To avoid polluting the namespace with conflicting identifiers, Haskell programmers can choose between explicit and qualified imports ; however, they are often reluctant to do so as:
+- Explicit import lists are inconvenient (they must be edited whenever the code changes). 
+- Module qualifiers can severely affect readability.
+
+The latter problem is often addressed by choosing one-letter qualifiers, which can only be understood by looking up their definition at the top of the file. It is an imperfect solution at best, and does not work well for operators ; for this reason, operators are frequently overloaded via lawless typeclasses for mere syntactic convenience.
+
+OCaml 3.12, released in 2011, introduced two syntactic contructs for locally opening modules: ``let open Module in …`` and ``Module.( … )``. In less than 10 years, this feature has gained widespread adoption throughout the ecosystem, bringing significant benefits in terms of readability and maintainability. For the reasons stated above, we believe Haskell should follow its example.
+
+Proposed Change Specification
+-----------------------------
+We allow qualified and non-qualified ``import`` statements at the beginning of any ``let``/``where``/``do``-block.
+
+For all constructs, the effect of such imports is limited to the scope the enclosing block. For ``let`` and ``where``, the imports also affect the corresponding expression. For example:
+::
+
+  let import Foo in <expr>
```

Comment:
The examples implies that `let` and `where` can contain `import` statement and other statements. However there is no discussion about the ordering of them. Is the following allowed:

```
let
    import A
    foo = bar
    import B
```

Or should `import` appear only in first position of the `let` / `where` block?

Same question for the `do` blocks. Does it affect all the block, or just the following statements? Can they appear everywhere in a `do` block or just at the beginning?

Reply:
I believe those should only be allowed at the beginning of a `let` or `where` block.

1) This is consistent with Haskell modules as a whole. I want to be as consistent as possible with existing restrictions.

2) Declarations in `let` and `where` clauses can be mutually recursive and appear in any order ; OTOH, in:

```
let
    import A
    foo = bar
    import B
```
it would feel quite unnatural to allow `bar` to use symbols from `B`.

Reply:
I have put emphasis on *beginning*.